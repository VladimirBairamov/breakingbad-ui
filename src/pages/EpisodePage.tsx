import clsx from 'clsx';
import Button from 'components/Button/Button';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useNavigate, useParams } from 'react-router-dom';
import { AppStore, RequestState } from 'store/common.types';
import { episodeGetOne } from 'store/episodes/episodes.actions';

const EpisodePage = () => {
  const { id } = useParams();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { currentEpisode, episodeByIdGetState } = useSelector((store: AppStore) => store.episodes);

  useEffect(() => {
    id && dispatch(episodeGetOne(+id));
  }, [dispatch, id]);

  if (episodeByIdGetState === RequestState.Loading || !currentEpisode) {
    return <>Loading...</>;
  }

  const formatTitle = (title: string) => {
    return (title[0].toUpperCase() + title.slice(1)).replace('_', ' ');
  };

  const onBackClick = () => {
    navigate(`/`);
  };

  return (
    <div className="bg-white shadow overflow-hidden sm:rounded-lg max-w-2xl mx-auto mt-16">
      <div className="px-4 py-2 flex justify-between">
        <h3 className="text-lg leading-6 font-medium text-gray-900 mt-3">Detailed information</h3>
        <Button onClick={onBackClick}>Go Back</Button>
      </div>
      <div className="border-t border-gray-200">
        <dl>
          {Object.entries(currentEpisode)
            .filter(([key]) => key !== 'characters')
            .map(([key, value], index) => (
              <div
                className={clsx(
                  'px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6',
                  (index % 2) - 1 ? 'bg-white' : 'bg-gray-50',
                )}
                key={key}
              >
                <dt className="text-sm font-medium text-gray-500">{formatTitle(key)}</dt>
                <dd className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2">{value}</dd>
              </div>
            ))}
          <div className="bg-white px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
            <dt className="text-sm font-medium text-gray-500">Characters</dt>
            <dd className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2">
              <ul className="border border-gray-200 rounded-md divide-y divide-gray-200">
                {currentEpisode.characters.map((character) => (
                  <li
                    className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
                    key={character}
                  >
                    <div className="w-0 flex-1 flex items-center">
                      <span className="ml-2 flex-1 w-0 truncate">{character}</span>
                    </div>
                    <div className="ml-4 flex-shrink-0">
                      <Link
                        to={`/character/${character}`}
                        className="font-medium text-indigo-600 hover:text-indigo-500"
                      >
                        Open
                      </Link>
                    </div>
                  </li>
                ))}
              </ul>
            </dd>
          </div>
        </dl>
      </div>
    </div>
  );
};

export default EpisodePage;
