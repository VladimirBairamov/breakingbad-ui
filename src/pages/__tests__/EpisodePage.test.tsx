import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { RequestState } from 'store/common.types';
import { useParams } from 'react-router-dom';
import EpisodePage from 'pages/EpisodePage';
jest.mock('react-redux');
jest.mock('react-router');

const mockData = {
  episodeByIdGetState: RequestState.Success,
  currentEpisode: {
    air_date: '03-02-2015',
    characters: [
      'Jimmy McGill',
      'Mike Erhmantraut',
      'Kim Wexler',
      'Howard Hamlin',
      'Chuck McGill',
      'Nacho Varga',
    ],
    episode: '5',
    episode_id: 67,
    season: '1',
    series: 'Better Call Saul',
    title: 'Alpine Shepherd Boy',
  },
};
beforeEach(() => {
  (useParams as any).mockImplementation(() => ({ id: '1' }));
  (useDispatch as any).mockImplementation(() => () => {});
});

test('EpisodePage must render loading', () => {
  (useSelector as any).mockImplementation(() => ({ episodeByIdGetState: RequestState.Loading }));
  const component = renderer.create(<EpisodePage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('EpisodePage must render data', async () => {
  (useSelector as any).mockImplementation(() => mockData);
  const component = renderer.create(<EpisodePage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
