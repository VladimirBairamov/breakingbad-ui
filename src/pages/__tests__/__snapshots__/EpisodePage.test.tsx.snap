// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodePage must render data 1`] = `
<div
  className="bg-white shadow overflow-hidden sm:rounded-lg max-w-2xl mx-auto mt-16"
>
  <div
    className="px-4 py-2 flex justify-between"
  >
    <h3
      className="text-lg leading-6 font-medium text-gray-900 mt-3"
    >
      Detailed information
    </h3>
    <button
      className="py-1 px-4 rounded my-2 mx-1 bg-gray-200 hover:bg-gray-400 text-black"
      onClick={[Function]}
    >
      Go Back
    </button>
  </div>
  <div
    className="border-t border-gray-200"
  >
    <dl>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-white"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Air date
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          03-02-2015
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-gray-50"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Episode
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          5
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-white"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Episode id
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          67
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-gray-50"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Season
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          1
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-white"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Series
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          Better Call Saul
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6 bg-gray-50"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Title
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          Alpine Shepherd Boy
        </dd>
      </div>
      <div
        className="bg-white px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Characters
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          <ul
            className="border border-gray-200 rounded-md divide-y divide-gray-200"
          >
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Jimmy McGill
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Mike Erhmantraut
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Kim Wexler
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Howard Hamlin
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Chuck McGill
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <span
                  className="ml-2 flex-1 w-0 truncate"
                >
                  Nacho Varga
                </span>
              </div>
              <div
                className="ml-4 flex-shrink-0"
              >
                <a
                  className="font-medium text-indigo-600 hover:text-indigo-500"
                  onClick={[Function]}
                >
                  Open
                </a>
              </div>
            </li>
          </ul>
        </dd>
      </div>
    </dl>
  </div>
</div>
`;

exports[`EpisodePage must render loading 1`] = `"Loading..."`;
